# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def add_two(num):
    return num + 2

add_two(4)

6

In [9]:
(lambda num: num + 2)(4)

6

In [11]:
(lambda num: num ** 2)(3)

9

In [12]:
a_list = [1,2,3,4,5]

[num ** 2 for num in a_list]

[1, 4, 9, 16, 25]

#### Saving to a Variable

In [14]:
square_nums = lambda x: x **2

In [16]:
square_nums(2)

4

In [17]:
square_nums(4)

16

In [18]:
square_nums(6)

36

#### Multiple Inputs

In [19]:
(lambda num1, num2, num3: num1+ num2+ num3)(4,5,6)

15

#### Passing a Lambda into a Function

In [21]:
def double_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] = a_list[i] * 2
    return a_list

double_numbers([1,2,3,4,5])

[2, 4, 6, 8, 10]

In [127]:
def half_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] = a_list[i] / 2
    return a_list

half_numbers([1,2,3,4,5])

[0.5, 1.0, 1.5, 2.0, 2.5]

In [130]:
def half_numbers(a_list):
    return [a_list[i] / 2 for i in (range(len(a_list)))]

half_numbers([1,2,3,4,5])

[0.5, 1.0, 1.5, 2.0, 2.5]

In [34]:
def increment_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] += 1   # a_list[i] = a_list[i] + 1
    return a_list

increment_numbers([1,2,3,4,5])

[2, 3, 4, 5, 6]

In [113]:
def increment_numbers(a_list):
    return [a_list[i] + 1 for i in range(len(a_list))]

increment_numbers([1,2,3,4,5])

[2, 3, 4, 5, 6]

In [32]:
def modify_numbers(a_list, f):
    for i in range(len(a_list)):
        a_list[i] = f(a_list[i])
    return a_list



In [36]:
# Double Numbers
modify_numbers([1,2,3,4,5], lambda x: x * 2)

[2, 4, 6, 8, 10]

In [38]:
# Half numbers
modify_numbers([1,2,3,4,5], lambda x: x / 2)

[0.5, 1.0, 1.5, 2.0, 2.5]

In [44]:
# Increment numbers
modify_numbers([1,2,3,4,5], lambda x: x + 1)

[2, 3, 4, 5, 6]

In [45]:
# Square our numbers, subtract 1, then add 4

modify_numbers([1,2,3,4,5], lambda x: x ** 2 - 1 + 4)

[4, 7, 12, 19, 28]

#### Returning a Lambda from a Function

In [62]:
lambda num1, num2: (num1 + num2 + 4) ** 2

<function __main__.<lambda>(num1, num2)>

In [64]:
lambda num1, num2: (num1 + num2 + 7) ** 3

<function __main__.<lambda>(num1, num2)>

In [71]:
def build_lambda(adder, power):
    return lambda num1, num2: (num1 + num2 + adder) ** power
#                                                 7     3
#                                                 4     2
# function for the lambda above 

In [77]:
build_lambda(7,3)
build_lambda(4,2)

<function __main__.build_lambda.<locals>.<lambda>(num1, num2)>

In [78]:
(build_lambda(7,3))(4,5)
#  adder = 7 power = 3 num1 = 4 num2 = 5

4096

#### If Statements within Lambdas

In [82]:
# square every other number in a list
a_list = [1,2,3,4,5]

modify_numbers(a_list, lambda x: x ** 2 if x % 2 == 1 else x)

[1, 2, 9, 4, 25]

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [84]:
def cubed(b_list, f):
    for i in range(len(b_list)):
        b_list[i] = f(b_list[i])
    return b_list

In [88]:
b_list = [1,2,3,4,5]

cubed(b_list, lambda x: x ** 3)

[1, 8, 27, 64, 125]

In [134]:
def cubed(c_list, f):
    return [f(x) for x in c_list]

In [135]:
c_list = [1,2,3,4,5,12,11]

cubed(c_list, lambda x: x ** 3)

[1, 8, 27, 64, 125, 1728, 1331]

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [120]:
a_list = [1,2,3,4,5]
print(list(map(lambda x: x * 2, a_list)))
print(list(map(lambda x: x * 3, a_list)))
print(list(map(lambda x: x * 4, a_list)))
print(list(map(lambda x: x ** 2, a_list)))

[2, 4, 6, 8, 10]
[3, 6, 9, 12, 15]
[4, 8, 12, 16, 20]
[1, 4, 9, 16, 25]


#### Using Lambda's with Map

In [121]:
print(list(map(lambda x: x ** 2, a_list)))

[1, 4, 9, 16, 25]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [126]:
a_list = [1,2,3,4,5]
print(list(map(lambda x: x * 2 - 1, a_list)))

[1, 3, 5, 7, 9]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [162]:
a_list = [1,2,3,4,5,6,7,8,9,10]

list(filter(lambda num: num % 2 == 0, a_list))

[2, 4, 6, 8, 10]

In [164]:
def is_odd(num):
    if num % 2 == 0:
        return True
    else:
        return False
list(filter(is_odd, a_list))

[2, 4, 6, 8, 10]

In [170]:
my_list = list('a,b,c') 
my_list

['a', ',', 'b', ',', 'c']

In [168]:
my_filtered_list = list(filter(lambda x: x != ',', my_list))

In [169]:
my_filtered_list

['a', 'b', 'c']

#### Using Lambda's with Filter()

In [172]:
a_list = [1,2,3,4,5,6,7,8,9,10]

list(filter(lambda num: num % 2 == 0, a_list))

[2, 4, 6, 8, 10]

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [185]:
import statistics

data = [220, 100, 190, 180, 250, 190, 240, 180, 140, 180, 190]
print(statistics.mean(data))

means = statistics.mean(data)

filtered_list = list(filter(lambda x: x < means, data))

print(filtered_list)


187.27272727272728
[100, 180, 180, 140, 180]


In [187]:
data = [220, 100, 190, 180, 250, 190, 240, 180, 140, 180, 190]

filtered_list = list(filter(lambda x: x < sum(data)/len(data), data))

print(filtered_list)

[100, 180, 180, 140, 180]


In [190]:
data = [220, 100, 190, 180, 250, 190, 240, 180, 140, 180, 190]

def mean1(data):
    total = 0 
    for i in range(len(data)):
        total += data[i]
    return total / len(data)

mean2 = mean1(data)
print(mean1(data))

filtered_list = list(filter(lambda x: x < mean2, data))

print(filtered_list)

187.27272727272728
[100, 180, 180, 140, 180]


In [184]:
data = [220, 100, 190, 180, 250, 190, 240, 180, 140, 180, 190]

def mean1(data):
    return sum(data) / len(data)

print(mean1(data))

187.27272727272728


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [196]:
from functools import reduce

# Use reduce to find the sum of a list

a_list = [1,2,3,4,5]

def add_nums(num1, num2):
    print(num1, num2)
    return num1 + num2

list_sum = reduce(add_nums, a_list)

1 2
3 3
6 4
10 5


In [198]:
print(list_sum)

15


In [200]:
def multiply_nums(num1, num2):
    return num1 * num2

reduce(multiply_nums, a_list)

120

#### Using Lambda's with Reduce()

In [201]:
reduce(lambda num1, num2: num1 * num2, a_list)

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [204]:
a_list = [1,2,3,4,5]
reduce(lambda num1, num2: num1 - num2, a_list)

-13

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [239]:
10 * 9 * 8 * 7 * 6 * 5 * 4 * 3 * 2 * 1

3628800

In [3]:
def factorial(number):
    # base case
    if number <= 1:
        return number
    
    return number * factorial(number - 1)

In [4]:
factorial(10)

3628800

#### Writing a Factorial Function

In [51]:
def factorial(number):
    return number if number <=1 else number * factorial(number-1)

In [52]:
factorial(10)

3628800

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [41]:
def factorial(number):
    return number if number <=1 else number - factorial(number-1)

In [47]:
print(factorial(5))

120


In [48]:
def sub(number):
    if number <= 1:
        print('sub(1) = 1')
        return number
    
    print(f'sub({number}) = {number} - sub({number - 1})')
    return number - sub(number - 1)

In [49]:
sub(5)

sub(5) = 5 - sub(4)
sub(4) = 4 - sub(3)
sub(3) = 3 - sub(2)
sub(2) = 2 - sub(1)
sub(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [60]:
my_nums = list(range(1,10))

In [58]:
def sqrt_nums(a_list):
    result = []
    for num in a_list:
        result.append(num ** 0.5)
    return result

In [62]:
sqrt_nums(my_nums)

[1.0,
 1.4142135623730951,
 1.7320508075688772,
 2.0,
 2.23606797749979,
 2.449489742783178,
 2.6457513110645907,
 2.8284271247461903,
 3.0]

In [64]:
def sqrt_nums(a_list):
    for num in a_list:
        yield num ** .5

In [71]:
my_nums_sqrt = sqrt_nums(my_nums)

In [74]:
next(my_nums_sqrt)

1.7320508075688772

In [76]:
my_nums_sqrt = list(sqrt_nums(my_nums))

In [77]:
my_nums_sqrt

[1.0,
 1.4142135623730951,
 1.7320508075688772,
 2.0,
 2.23606797749979,
 2.449489742783178,
 2.6457513110645907,
 2.8284271247461903,
 3.0]

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [2]:
def sqrt_nums(a, b, c = 1):
       while b >= a:
        yield b ** 2
        b -= c
        
for i in sqrt_nums(0,b = 3):
    print(i)

9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [214]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filtered_places = list(filter(lambda x: x.strip() != "", places))

print(filtered_places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [112]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda x: x.split()[-1])

print(author)

['Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'Victor aNisimov', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [107]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

print(list(map(lambda x: (x[0], (9/5)*x[1] + 32), places)))

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


In [111]:
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

def place():
    return [(x[0], (9/5)*x[1] + 32) for x in places]

place()

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [10]:
def fib(n):
    n = abs(n)
    
    if n == 0:
        return 1
    elif n == 1:
        return 1
    else: return fib(n-1) + fib(n-2)
    
for i in range(0,6):
    print('fib(', i, ') =', fib(i))

fib( 0 ) = 1
fib( 1 ) = 1
fib( 2 ) = 2
fib( 3 ) = 3
fib( 4 ) = 5
fib( 5 ) = 8


In [12]:
def fib(number):
    if number <= 1:
        return number
    else:
        return (fib(number-1) + fib(number-2))
num = 8
for i in range(num):
    print(fib(i))

0
1
1
2
3
5
8
13
